schema/dmt/build.go

```diff
@@ -0,0 +1,71 @@
+package schemadmt
+
+import (
+	"fmt"
+
+	"github.com/ipld/go-ipld-prime/schema"
+)
+
+func Load(ts *schema.TypeSystem, node *Schema) error {
```

Comment:
So this function is the big crux where things connect -- it mutates the given `*schema.TypeSystem` pointer.

Just gonna talk my way through what the design approach is here: The way this function is approaching things is: we've got this `schemadmt` package, and we're still expecting the `schema` package to expose the `Spawn*` functions.  The dependency arrow order is `schemadmt -> schema`, and that's it.  Users would have to import the `schemadmt` package, use it to parse and load the data, and use this function to compile it.  And we've left all the compilation validation logic in the `schema` package.  (Implicitly, it is still very possible right now to create `schema.*` values that are nonsensical, and the interfaces don't stop you at all.)  And the amount of mostly-repetitive hand-wrought golang types and accessor methods in the `schema` package remains high.  (But, in this approach, since it's using bindnode instead of my heavy-duty (and I don't mean that in a good way) codegen, we don't have any of those accessor methods on the dmt types, so the repetitivity level, as it manifests in SLOC, is a lot lower.)  Okay.  Cool.

Reply:
There are a couple small things I'd probably change, just at the level of how we call this:

- I'd call this `Compile`.  It's going from DMT (something pretty legible, transcodable, etc) to the form _used_ to do work, which is not reversable back to DMT.  An operation that goes from legibility, irreversibly, to illegible-but-useful sounds like a "compile" operation to me.

- Not sure if thrilled with the mutate-the-argument strategy.  My first inclination would be to have it look more like a constructor and return a new `schema.TypeSystem`.  I don't think we have a lot of reasons to think of appending a TypeSystem repeatedly.  (Like, even if it's currently mechanically possible, A) I don't think it's a good API contract for TypeSystem to be treated as mutable, and B) just... why?  It's something one would only be able to do by writing Interesting golang code.  When we think of how we'll have schema-driven tooling work purely by APIs (think: IPLD Explorer with a Schema parameter), none of those Interesting ways one could compose golang code will be relevant, so istm it's somewhere between not useful and mildly bad for us to have API features that won't make sense when we go data-driven.)

Reply:
I totally made up the name Load, fully expecting we'd replace it :) Compile sounds fine. I also thought of others like Accumulate or Spawn.

Regarding append to vs return a new TypeSystem: I thought I'd leave the API being more flexible, just like Accumulate. I imagine it could be reasonable to use a combination of two schemas at once, for example. If that's something we never want to allow, then SGTM to return a new typesystem.

Reply:
Your understanding of the dependency graph is right. That's what I arrived at with my paper sketches some weeks ago. When we end up with a fully end-to-end Parse API that takes a file and gives you a TypeSystem, that will probably live in schemadmt, or a package that imports schemadmt.

And you're right that there's still glue code to walk the "AST" nodes and spawn schema types. But like you say, this is pretty painless with "just Go types" :)